app/helpers/samples_helper.rb

```diff
@@ -619,7 +619,50 @@ def augment_taxon_list_with_sample_count(taxon_list, samples)
                       .map { |r| [r.tax_id, r.sample_count] }
                       .to_h
     taxon_list.each do |taxon|
-      taxon["sample_count"] = counts_by_taxid[taxon["taxid"]]
+      taxon["sample_count"] = counts_by_taxid[taxon["taxid"]] || 0
+    end
+    taxon_list
+  end
+
+  # For each taxon, count how many samples have contigs for that taxon.
+  # Add these counts to the taxon objects.
+  def augment_taxon_list_with_sample_count_contigs(taxon_list, samples)
+    get_tax_ids_by_level = lambda do |level|
+      taxon_list.select { |taxon| taxon["level"] == level }.map { |taxon| taxon["taxid"] }
+    end
+
+    # Separate genus and species taxids. We query them separately to try to get some extra performance.
+    species_tax_ids = get_tax_ids_by_level.call("species")
+    genus_tax_ids = get_tax_ids_by_level.call("genus")
+
+    pipeline_run_ids = get_succeeded_pipeline_runs_for_samples(samples).pluck(:id)
+
+    # We maintain a hash of taxid => array of pipeline run ids. We de-dupe the pipeline run ids at the very end.
+    pipeline_runs_by_taxid = Hash.new { |h, k| h[k] = [] } # Keys are auto-initialized to the empty array.
+
+    process_tax_ids = lambda do |level, type|
```

Comment:
note: I would not abuse of lambdas... seems like a a normal method would be good here.

Reply:
The method uses `species_tax_ids` and `genus_tax_ids`, which are from the function scope. Is there a downside to using lambdas? I typically only use them for nested functions.

Reply:
IMO, just readability, since they are in the middle of another function. It's like code interleaved with functions, makes it hard to read.

Reply:
That makes sense. I can make it a convention to move lambdas to the top of the function, instead of interleaving them.

I tend to use lambdas for cases where the lambda code is specific to the containing function. You could extract the lambda code outside the parent function, but that can also hurt readability because now the fact that the lambda is intended to only be used within the parent function is less clear.